File: 678.valid-parenthesis-string.cpp
/*
 * @lc app=leetcode id=678 lang=cpp
 *
 * [678] Valid Parenthesis String
 */
#include <string>
using namespace std;
// @lc code=start
class Solution
{
public:
    bool checkValidString(string s)
    {
        int muteki_symbol = 0;
        int trans_to_left_symbol = 0;
        int rest_min_left = 0;
        for (auto c : s)
        {
            if (c == '(')
            {
                ++rest_min_left;
            }
            else if (c == ')')
            {
                if (rest_min_left > 0)
                {
                    --rest_min_left;
                }
                else if (muteki_symbol > 0)
                {
                    --muteki_symbol;
                }
                else if (trans_to_left_symbol > 0)
                {
                    --trans_to_left_symbol;
                    ++muteki_symbol;
                }
                else
                {
                    return false;
                }
            }
            else
            {
                if (rest_min_left > 0)
                {
                    --rest_min_left;
                    ++trans_to_left_symbol;
                }
                else
                {
                    ++muteki_symbol;
                }
            }
        }
        return rest_min_left == 0;
    }
};

// @lc code=end

File: py/678.valid-parenthesis-string.py
#
# @lc app=leetcode id=678 lang=python3
#
# [678] Valid Parenthesis String
#

# @lc code=start
class Solution:
    def checkValidString(self, s: str) -> bool:
        muteki_symbol = 0
        trans_to_left_symbol = 0
        rest_min_left = 0
        for c in s:
            if c == '(':
                rest_min_left += 1
            elif c == ')':
                if rest_min_left > 0:
                    rest_min_left -= 1
                elif muteki_symbol > 0:
                    muteki_symbol -= 1
                elif trans_to_left_symbol > 0:
                    trans_to_left_symbol -= 1
                    muteki_symbol += 1
                else:
                    return False
            else:
                if rest_min_left > 0:
                    rest_min_left -= 1
                    trans_to_left_symbol += 1
                else:
                    muteki_symbol += 1
        return rest_min_left == 0
# @lc code=end
